packages/block-library/src/navigation/edit.js

```diff
@@ -119,132 +84,247 @@ function Navigation( {
-			</TextColor>
+			</Block.nav>
 		</>
 	);
 }
 
-export default compose( [
-	withSelect( ( select, { clientId } ) => {
-		const innerBlocks = select( 'core/block-editor' ).getBlocks( clientId );
+export const withInspectorControls = createHigherOrderComponent(
+	( BlockEdit ) => ( props ) => {
+		return (
+			<>
+				<BlockEdit { ...props } />
+				<InspectorControls>
+					<PanelBody title={ __( 'Display settings' ) }>
+						<ToggleControl
+							checked={ props.attributes.showSubmenuIcon }
+							onChange={ ( value ) => {
+								props.setAttributes( {
+									showSubmenuIcon: value,
+								} );
+							} }
+							label={ __( 'Show submenu indicator icons' ) }
+						/>
+					</PanelBody>
+				</InspectorControls>
+			</>
+		);
+	},
+	'withInspectorControls'
+);
+
+addFilter(
+	'navigation.BlockEdit',
+	'core/block-library/navigation/with-inspector-controls',
+	withInspectorControls
+);
+
+export const withBlockControls = createHigherOrderComponent(
+	( BlockEdit ) => ( props ) => {
+		const { attributes, setAttributes } = props;
+		//
+		// HANDLERS
+		//
+		function handleItemsAlignment( align ) {
+			return () => {
+				const itemsJustification =
+					attributes.itemsJustification === align ? undefined : align;
+				setAttributes( {
+					itemsJustification,
+				} );
+			};
+		}
+		return (
+			<>
+				<BlockEdit { ...props } />
+				<BlockControls>
+					<ToolbarGroup
+						icon={
+							attributes.itemsJustification
+								? navIcons[
+										`justify${ upperFirst(
+											attributes.itemsJustification
+										) }Icon`
+								  ]
+								: navIcons.justifyLeftIcon
+						}
+						label={ __( 'Change items justification' ) }
+						isCollapsed
+						controls={ [
+							{
+								icon: navIcons.justifyLeftIcon,
+								title: __( 'Justify items left' ),
+								isActive:
+									'left' === attributes.itemsJustification,
+								onClick: handleItemsAlignment( 'left' ),
+							},
+							{
+								icon: navIcons.justifyCenterIcon,
+								title: __( 'Justify items center' ),
+								isActive:
+									'center' === attributes.itemsJustification,
+								onClick: handleItemsAlignment( 'center' ),
+							},
+							{
+								icon: navIcons.justifyRightIcon,
+								title: __( 'Justify items right' ),
+								isActive:
+									'right' === attributes.itemsJustification,
+								onClick: handleItemsAlignment( 'right' ),
+							},
+						] }
+					/>
+				</BlockControls>
+			</>
+		);
+	},
+	'withBlockControls'
+);
+
+addFilter(
+	'navigation.BlockEdit',
+	'core/block-library/navigation/with-block-controls',
+	withBlockControls
+);
+
+export const withListView = createHigherOrderComponent(
+	( BlockEdit ) => ( props ) => {
+		const { clientId } = props;
+		const { navigatorToolbarButton, navigatorModal } = useBlockNavigator(
+			clientId
+		);
+
+		return (
+			<>
+				<BlockEdit { ...props } />
+				<BlockControls>
+					<ToolbarGroup>{ navigatorToolbarButton }</ToolbarGroup>
```

Comment:
Hopefully it'll be possible to replace this with https://github.com/WordPress/gutenberg/pull/25034